extra/add_tenant_token.sh

```diff
@@ -0,0 +1,125 @@
+#!/bin/bash
+set -e
+
+if [ -n "$DEBUG" ]; then
+  set -x
+fi
+
+REQUIREMENTS=( "jq" "debugfs" "sha256sum" "parted" )
+HOSTED_MENDER_URL="https://www.menderlab.xyz"
+
+function mender_tt {
+  local mender_conf_file="/etc/mender/mender.conf"
+  local filename=$1
+  local tt=$2
+
+  local tmp_dir=$(mktemp -d)
+  local jq_argument=$(printf ".TenantToken = \"%s\" | .ServerURL = \"%s\"" "$tt" "$HOSTED_MENDER_URL")
+  local tmpfile=$(mktemp)
+
+  tar xf "$filename" -C "$tmp_dir"
+  tar xf "$tmp_dir"/data/0000.tar.gz -C "$tmp_dir"/data/
+
+  local foundext4=$(find "$tmp_dir" -iname "*.ext4")
+  local ext4filename=$(basename "$foundext4")
+
+  # copy mender.conf file from image
+  debugfs -R "dump $mender_conf_file $tmp_dir/mender.conf" $foundext4
+
+  # update JSON with tenant key and hosted mender ServerURL
+  jq "$jq_argument" "$tmp_dir/mender.conf" > "$tmpfile"
+
+  # create debugfs script to copy over new mender.conf file
+  echo "cd /etc/mender/" > /tmp/tt_script
+  echo "rm mender.conf" >> /tmp/tt_script
+  echo "write $tmpfile mender.conf" >> /tmp/tt_script
+  echo "close" >> /tmp/tt_script
+
+  echo "New config. file is:"
+  cat "$tmpfile"
+
+  # write back to image
+  debugfs -w -f "/tmp/tt_script" "$foundext4"
+
+  local newsha=$(sha256sum "$foundext4" | head -c 64)
+  sed -i "/$ext4filename/d" "$tmp_dir/manifest"
+  echo "$newsha  data/0000/$ext4filename" >> "$tmp_dir/manifest"
+
+  # repackage image
+  (cd "$tmp_dir"/data && tar czf 0000.tar.gz "$ext4filename")
+
+  # re-create mender artifact
+  local new_mender_artifact=$(pwd -P)/$(basename "$filename" .mender)-"${tt: -6}"".mender"
+  (cd "$tmp_dir"/ && tar cf "$new_mender_artifact" version manifest header.tar.gz data/0000.tar.gz)
+  echo
+  echo "Mender artifact with specified tenant_token created with filename: $new_mender_artifact"
+
+  rm -rf "$tmp_dir"
+  rm "$tmpfile"
+}
+
+function sdimg_tt {
+  local mender_conf_file="/etc/mender/mender.conf"
+  local filename=$1
+  local tt=$2
+
+  if [ "$EUID" -ne 0 ]; then
+    echo "Please run as root"; exit 1
+  fi
+
+  local mount_dir=$(mktemp -d)
+
+  # ugly way to determine start sectors of sdimg
+  local start_sectors=$(sudo parted "$filename" "unit s print" | grep -i '^[[:blank:]]\{1,\}[0-9]' | awk '{ print $2}' | tr -d 's')
+  read -a start_sectors_array <<< $start_sectors
+
+  local tempfile=$(mktemp)
+  local jq_argument=$(printf ".TenantToken = \"%s\" | .ServerURL = \"%s\"" "$tt" "$HOSTED_MENDER_URL")
+
+  # mount every partition, and copy over the mender.conf file
+  for s in "${start_sectors_array[@]}"
+  do
+    sudo mount -o loop,offset=$((s * 512)) "$filename" "$mount_dir"
```

Comment:
Why not `debugfs` here as well? I assume this will run on a SaaS server, and I really don't think sudo is a good idea there.

Reply:
Oh right, you will have to copy out the file in that case. Bah... Well, it's up to you. This is a stop gap solution anyway, since Marcin is working on proper udisks support.